File: java/src/main/java/br/com/emendes/problems/FirstMissingPositive.java
package br.com.emendes.problems;

import java.util.HashSet;
import java.util.Set;

/**
 * Dado um array de inteiros não ordenado nums.
 * Retorne o menor inteiro positivo ausente.<br><br>
 * Voçê deve implementar o algoritmo com complexidade de tempo O(n) e complexidade de espaço em O(1).<br><br>
 * Restrições:<br>
 * 1 <= nums.length <= 10⁵<br>
 * -2³¹ <= nums[i] <= 2³¹ - 1
 */
public class FirstMissingPositive {

  public int firstMissingPositive(int[] nums) {
    for (int i = 0; i < nums.length; i++) {
      while (nums[i] > 0 && nums[i] <= nums.length && nums[nums[i] - 1] != nums[i]) {
        swap(i, nums[i] - 1, nums);
      }
    }
    for (int i = 0; i < nums.length; i++) {
      if (nums[i] != i + 1) return i + 1;
    }
    return nums.length + 1;
  }

  private void swap(int i1, int i2, int[] arr) {
    int temp = arr[i1];
    arr[i1] = arr[i2];
    arr[i2] = temp;
  }

  public int solution2(int[] nums) {
    int length = nums.length;
    int[] answer = new int[length + 1];

    for (int i = 0; i < length; i++) {
      if (nums[i] > 0 && nums[i] <= length) {
        answer[nums[i]] = 1;
      }
    }

    length = length + 1;
    for (int i = 1; i < length; i++) {
      if (answer[i] == 0) return i;
    }

    return length;
  }

  // Complexidade de tempo O(n)
  // Complexidade de espaço O(n)
  public int solution1(int[] nums) {
    Set<Integer> remainder = new HashSet<>();
    int smallest = 1;
    for (int num : nums) {
      if (num >= smallest) {
        if (num == smallest) {
          smallest++;
        } else {
          remainder.add(num);
        }
      }
    }

    while (remainder.contains(smallest)) {
      smallest++;
    }

    return smallest;
  }

}

File: javascript/problems/FirstMissingPositive.js
/**
 * Dado um array de inteiros nums não ordenado. Retorne o menor inteiro positivo que não está presente em nums.
 *
 * Você deve implementar um algoritmo que executa em com complexidade de tempo de O(n) e
 * complexidade de espaço O(1).
 *
 * Restrições:
 *
 * 1 <= nums.length < 10⁵
 *
 * -2³¹ <= nums[i] <= 2³¹ - 1
 *
 * @param {number[]} nums
 * @return {number}
 */
const firstMissingPositive = function (nums) {
  for (let index = 0; index < nums.length; index++) {
    if (nums[index] < 0) nums[index] = 0;
  }
  const maxValue = nums.length;
  for (const key in nums) {
    const element = Math.abs(nums[key]);
    if (element >= 1 && element <= maxValue) {
      nums[element - 1] = nums[element - 1] === 0 ? -element : Math.abs(nums[element - 1]) * -1;
    }
  }
  let index = 1;
  while (index <= maxValue && nums[index - 1] < 0) {
    index++;
  }
  return index;
};
